server/src/future.rs

```diff
@@ -241,3 +241,43 @@ impl Future for ShutdownWaiter {
 		}
 	}
 }
+
+/// A `Future` that resolves once the server has been requested to be stopped.
+#[derive(Debug)]
+pub(crate) struct ServerStopped(Weak<MonitorInner>);
+
+impl Future for ServerStopped {
+	type Output = ();
+
+	fn poll(self: Pin<&mut Self>, _cx: &mut Context) -> Poll<Self::Output> {
+		if let Some(arc) = Weak::upgrade(&self.0) {
+			// We expect this method to be polled frequently, skipping an iteration isn't problematic, so relaxed
+			// ordering is optimal.
+			if !arc.shutdown_requested.load(Ordering::Relaxed) {
+				return Poll::Pending;
```

Comment:
If this returns `Pending` without doing anything with the Context/Waker (and calling it when it's ready to continue) the future might not be polled again?

Reply:
.. so perhaps we need to do what ShutdownWaiter does here and call `arc.waker.register(cx.waker())` just before we return `Poll::Pending`?

Reply:
So I was mulling over this more:

By the looks of it, the waker is only woken when the `StopMonitor` thing is dropped. It isn't woken when we flip the `shutdown_requested` bool to true. 

So, At the moment I don't think there's a way to have a future which will return when shutdown_requested becomes true. I think we'd need to add another waker which was `.wake()`d once the bool has definitely become true in order to do this.

Without more digging I don't know whether we need this or not though, and what the benefit of knowing when shutdown ahs been requested is :)

Reply:
this is because we need to trigger the graceful shutdown of hyper, and shutdown waiter waits for all background tasks to complete.

I need to double-check again why the ShutdownWaiter didn't work for this again, I think it's because it's not triggered because the strong count on the Arc is bigger than 1 all the time and because it's cloned in hyper the service_handle